### PR TITLE
gcs: tl_dfu: retry initial connection

### LINKED ISSUE #888: Upgrader failed to connect to loader on windows 7
From user `crashnburn0` on irc:

```
19:57:44[DEBUG]"UpgradeAssistant: in step 4"
19:57:46[DEBUG]Removing  8352 16731 1025 "52FF77065077525137130787+BL" "CopterControl" "openpilot.org"
19:57:46[DEBUG]usbmonitor detection cycle complete.
19:57:46[DEBUG]"Bootloader disconnection detected"
19:57:48[DEBUG]Adding  8352 16731 3 "52FF77065077525137130787+UP" "CopterControl3D" "openpilot.org"
19:57:48[DEBUG]usbmonitor detection cycle complete.
19:57:48[DEBUG]"Could not open USB port"
19:57:48[DEBUG]"Could not open coms with the bootloader"
19:58:09[DEBUG]"Upgrade assistant failed: Unable to enter legacy upgrade tool on board!!"
```

### PR DESCRIPTION
**Pull request by mlyle**
For consideration in tanto.

Needs careful code audit and test.

fixes #888

### CODE REVIEW

File: ground/gcs/src/plugins/uploader/tl_dfu.cpp [hunk 1]
```diff
@@ -453,12 +453,15 @@ device DFUObject::findCapabilities()
   */
 bool DFUObject::OpenBootloaderComs(USBPortInfo port)
 {
+    int retries=3;
```
Comment (tracernz) [inline ground/gcs/src/plugins/uploader/tl_dfu.cpp]: Spaces

Comment (mlyle) [inline ground/gcs/src/plugins/uploader/tl_dfu.cpp]: woops

**Comment by mlyle** [comment]
Not really possible to test this...

I've skimmed it and I'm happy with the code.

File: ground/gcs/src/plugins/uploader/tl_dfu.cpp [hunk 2]
```diff
@@ -469,11 +472,19 @@ bool DFUObject::OpenBootloaderComs(USBPortInfo port)
         AbortOperation();
         if(!EnterDFU()) {
             TL_DFU_QXTLOG_DEBUG(QString("Could not process enterDFU command"));
+            if ((retries--) > 0) {
```
Comment (tracernz) [inline ground/gcs/src/plugins/uploader/tl_dfu.cpp]: Qt style: "Use curly braces only when the body of a conditional statement contains more than one line:" Applies below 2 more times too.

Comment (mlyle) [inline ground/gcs/src/plugins/uploader/tl_dfu.cpp]: :( i really don't like that style.  pretty please no?

Comment (mlyle) [inline ground/gcs/src/plugins/uploader/tl_dfu.cpp]: Basically how I see the pro/con--  skipping the brackets saves some screen space.  But it makes it easy to group statements wrong accidentally and have really unexpected behavior.

Comment (tracernz) [inline ground/gcs/src/plugins/uploader/tl_dfu.cpp]: Both flight and GCS style call for it... Flight/Linux:

```
Do not unnecessarily use braces where a single statement will do.

    if (condition)
        action();

and

    if (condition)
        do_this();
    else
        do_that();
```

If you really wanna do it I guess we can.

**Comment by tracernz** [comment]
Worst case it fails 3 times instead of 1, normal case it makes no difference and best case it helps somebody.

**Pull request merged**